tensorboard/webapp/runs/store/utils.ts

```diff
@@ -41,8 +47,59 @@ export function groupRuns(
       break;
 
     case GroupByKey.REGEX:
-      throw new Error('Not implemented');
+      if (!groupBy.regexString) {
+        //TODO(japie1235813): propagate invalidity of regex string to user more gracefully
+        break;
+      }
+      let regExp: RegExp;
+
+      // TODO(japie1235813): add additonal `\` to convert string to regex, which
+      // makes `new RegExp()` construct properly
+      // For example, convert `foo\d+bar` to `foo\\d+bar`
+
+      try {
+        regExp = new RegExp(groupBy.regexString);
+      } catch (e) {
+        //TODO(japie1235813): propagate invalidity of regex string to user more gracefully
+        break;
+      }
+
+      // Checks if there is capture group in regex.
+      let isCaptureGroup = false;
+
+      for (const run of runs) {
+        let matches = run.name.match(regExp);
+        if (matches) {
+          if (matches.length > 1) {
+            matches = matches.slice(1);
+            isCaptureGroup = true;
+          }
+          const id = JSON.stringify(matches);
+          const runs = runGroups[id] || [];
+          runs.push(run);
+          runGroups[id] = runs;
+        } else {
+          nonMatches.push(run);
+        }
+      }
+      runGroup.nonMatches = nonMatches;
+
+      if (!isCaptureGroup) {
+        // No capture group in regex string. Groups all the matched runs together.
+        const matchedRuns = [];
+        runGroups = {};
+        for (const run of runs) {
+          let matches = run.name.match(regExp);
+          if (matches) {
+            matchedRuns.push(run);
+            delete runGroups[run.id];
+          }
+        }
+        runGroups['matches'] = matchedRuns;
+      }
+      break;
```

Comment:
Currently there are two for-loops for `runs` in case the regex string does not have the capture group. How about instead, we do:
```ts
for (const run of runs) {
  const matches = run.name.match(regExp);
  if (matches) {
    const hasCapturingGroup = matches.length > 1;
    // In case regex string does not have a capturing group, we use pseudo group id of `pseudo_group`.
    const id = hasCapturingGroup ? JSON.stringify(matches.slice(1)) : 'pseudo_group';
    const runs = runGroups[id] || [];
    runs.push(run);
    runGroups[id] = runs;
  } else {
    nonMatches.push(run);
  }
}
```